lib/system/nuttx/irq.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2018, Pinecone Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file	nuttx/irq.c
+ * @brief	NuttX libmetal irq definitions.
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/alloc.h>
+#include <nuttx/irq.h>
+
+/** IRQ handler descriptor structure */
+struct metal_irq_hddesc {
+	metal_irq_handler hd;     /**< irq handler */
+	void *drv_id;             /**< id to identify the driver
+	                               of the irq handler */
+};
+
+static int metal_irq_isr(int irq, void *context, void *arg);
+
+int metal_irq_register(int irq,
+		       metal_irq_handler hd,
+		       struct metal_device *dev,
+		       void *drv_id)
```

Comment:
could you take a look at this pull request: https://github.com/OpenAMP/libmetal/pull/72?
It will introduce a change in the IRQ implementation, which will introduce interrupt controller.
Here is the branch for the change: https://github.com/wjliang/libmetal/tree/master-soft-irq
Please check if it will have a big impact in your implementation.

Thanks,
Wendy

Reply:
Ok, I will study the patch in the weekend and summary the impact.

Reply:
It's simple to port to the new desgin, but we need improve the desgin to avoid the handler array duplication if RTOS already have the similiar irq dispatch table.